xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java

```diff
@@ -0,0 +1,224 @@
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    mtlsCommonTest(upstreamTlsContext);
+  }
+
+  /** mTLS - client auth enabled then update server certs to untrusted. */
+  @Test
+  public void mtlsClientServer_changeServerContext_expectException() throws IOException {
+    UpstreamTlsContext upstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildUpstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.CLIENT_KEY_FILE,
+            SecretVolumeSslContextProviderTest.CLIENT_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    XdsClient.ListenerWatcher listenerWatcher = mtlsCommonTest(upstreamTlsContext);
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            BAD_SERVER_KEY_FILE,
+            BAD_SERVER_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    createListenerUpdate(server.getPort(), downstreamTlsContext, listenerWatcher);
+    try {
+      buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+      fail("exception expected");
+    } catch (StatusRuntimeException sre) {
+      assertThat(sre).hasCauseThat().isInstanceOf(SSLHandshakeException.class);
+      assertThat(sre).hasCauseThat().hasMessageThat().isEqualTo("General OpenSslEngine problem");
+    }
+  }
+
+  private XdsClient.ListenerWatcher mtlsCommonTest(UpstreamTlsContext upstreamTlsContext)
+      throws IOException {
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.SERVER_1_KEY_FILE,
+            SecretVolumeSslContextProviderTest.SERVER_1_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+
+    XdsClient.ListenerWatcher listenerWatcher = getXdsServer(downstreamTlsContext);
+    buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+    return listenerWatcher;
+  }
+
+  private XdsClient.ListenerWatcher getXdsServer(DownstreamTlsContext downstreamTlsContext)
+      throws IOException {
+    int port = findFreePort();
+    XdsServerBuilder builder =
+        XdsServerBuilder.forPort(port).addService(new SimpleServiceImpl());
+    final XdsClientWrapperForServerSds xdsClientWrapperForServerSds =
+        createXdsClientWrapperForServerSds(port, downstreamTlsContext);
+    SdsProtocolNegotiators.ServerSdsProtocolNegotiator serverSdsProtocolNegotiator =
+        new SdsProtocolNegotiators.ServerSdsProtocolNegotiator(xdsClientWrapperForServerSds);
+    server = cleanupRule.register(builder.buildServer(serverSdsProtocolNegotiator)).start();
+    return xdsClientWrapperForServerSds.getListenerWatcher();
+  }
+
+  /** Creates XdsClientWrapperForServerSds: also used by other classes. */
+  public static XdsClientWrapperForServerSds createXdsClientWrapperForServerSds(
```

Comment:
should probably moved to util class.

Reply:
ok

Reply:
Moved to `XdsClientWrapperForServerSdsTest` since it cannot be in `CommonTlsContextTestsUtil` due to different package and I didn't see the need to create a  new util class here just for this